File: 18#4Sum.py
class Solution(object):
    def threeSum(self, nums,targetR):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        size = len(nums)
        nums.sort()
        for i in range(0,size-2):
            target = nums[i]
            left = i+1
            right = size-1
            while left<right:
                if nums[left]+nums[right]+target>targetR:
                    right -= 1
                elif nums[left]+nums[right]+target<targetR:
                    left += 1
                else:
                    if [nums[i],nums[left],nums[right]] not in result:
                        result.append([nums[i],nums[left],nums[right]])
                    left += 1
        return result

    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        for i in range(len(nums)):
            if i<len(nums)-2:
                tempR = self.threeSum(nums[i+1:],target-nums[i])
                if len(tempR)>0:
                    for li in tempR:
                        li.append(nums[i])
                        if li not in res:
                            res.append(li)
        return res

if __name__ == '__main__':
    so = Solution()
    so.fourSum([1,0,-1,0,-2,2],0)

File: 18#4Sum.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

int compInc(const void *a, const void *b){
    return *(int *)a - *(int *)b;
}
int** threeSum(int* nums, int numsSize, int* returnSize，int tar) {
    int count = 0;
    int** result = (int**)malloc(sizeof(int*)*(numsSize*(numsSize-1)*(numsSize-2))/6);
    qsort(nums,numsSize,sizeof(nums[0]),compInc);

    for(int i=0;i<numsSize-1;i++){
        int target = nums[i];
        if(i>0&&nums[i]==nums[i-1]) continue;
        int left = i+1;
        int right = numsSize-1;
        while(left<right){
            if(nums[left]+nums[right]+target>tar){
                right--;
                continue;
            }
            if(nums[left]+nums[right]+target<tar){
                left++;
                continue;
            }
            if(nums[left]+nums[right]+target==tar){
                result[count] = (int*)malloc(sizeof(int)*3);
                result[count][0] = nums[i];
                result[count][1] = nums[left];
                result[count][2] = nums[right];
                count ++;
                left++;
                continue;
            }
        }
    }
    *returnSize = count;
    printf("%d\n",*returnSize);
    for(int j=0;j<count;j++){
        for(int i=0;i<3;i++){
            printf("%d,",result[j][i]);
        }
        printf("\n");
    }

    return result;
}

int** fourSum(int* nums, int numsSize, int target, int* returnSize) {
  qsort(nums,numsSize,sizeof(nums[0]),compInc);
  /*
  待补充
  */
  
}

int main(int argc, char const *argv[]) {
  /* code */

  return 0;
}
